libraries/hotstuff/qc_chain.cpp

```diff
@@ -388,81 +350,74 @@ namespace eosio::hotstuff {
-
-         // if a finalizer has already aggregated a vote signature for the current QC, just discard this vote
-
+      if (!quorum_met) {
          const auto& finalizers = _pacemaker->get_finalizer_set().finalizers;
          for (size_t i=0; i<finalizers.size(); ++i)
-            if (finalizers[i].public_key == vote.finalizer_key)
-               if (finalizer_set.test(i))
-                  return;
-
-         if (finalizer_set.any())
-            _current_qc.set_active_agg_sig(fc::crypto::blslib::aggregate({_current_qc.get_active_agg_sig(), vote.sig }));
-         else
-            _current_qc.set_active_agg_sig(vote.sig);
-         fc_tlog(_logger, " === update bitset ${value} ${finalizer_key}", ("value", _current_qc.get_active_finalizers_string())("finalizer_key", vote.finalizer_key));
-         _current_qc.set_active_finalizers(update_bitset(finalizer_set, vote.finalizer_key));
-
-         quorum_met = is_quorum_met(_current_qc, *p);
-
-         if (quorum_met){
-
-            fc_dlog(_logger, " === ${id} quorum met on #${block_num} ${phase_counter} ${proposal_id} ",
-                           ("block_num", p->block_num())
-                           ("phase_counter", p->phase_counter)
-                           ("proposal_id", vote.proposal_id)
-                           ("id", _id));
-
-            _current_qc.set_quorum_met();
-
-            //fc_tlog(_logger, " === update_high_qc : _current_qc ===");
-            update_high_qc(_current_qc);
-
-            //check for leader change
-            leader_rotation_check();
-
-            //if we're operating in event-driven mode and the proposal hasn't reached the decide phase yet
-            if (_chained_mode == false && p->phase_counter < 3) {
-               fc_tlog(_logger, " === ${id} phase increment on proposal ${proposal_id}", ("proposal_id", vote.proposal_id)("id", _id));
-               hs_proposal_message proposal_candidate;
-
-               if (_pending_proposal_block.empty())
-                  proposal_candidate = new_proposal_candidate( p->block_id, p->phase_counter + 1 );
-               else
-                  proposal_candidate = new_proposal_candidate( _pending_proposal_block, 0 );
-
-               reset_qc(proposal_candidate.proposal_id);
-               fc_tlog(_logger, " === ${id} setting _pending_proposal_block to null (process_vote)", ("id", _id));
-
-               _pending_proposal_block = {};
-               _b_leaf = proposal_candidate.proposal_id;
-
-               //todo : asynchronous?
-               //write_state(_liveness_state_file , _liveness_state);
-
-               send_hs_proposal_msg( std::nullopt, proposal_candidate );
-
-               fc_tlog(_logger, " === ${id} _b_leaf updated (process_vote): ${proposal_id}", ("proposal_id", proposal_candidate.proposal_id)("id", _id));
+            if (finalizers[i].public_key == vote.finalizer_key) {
+               digest_type digest = p->get_proposal_digest();
```

Comment:
This should be called once outside of the loop. Also I think it should be renamed to `calculate_proposal_digest()` as it is not a getter. We may want to consider only creating the proposal digest once and caching it in the `hs_proposal_message` or somewhere else.

Reply:
done.